### PR TITLE
Add large runtime resource directory to lfs, and expose it to the Docker build.

### PR DESCRIPTION
**Pull request by cmnbroad**
This adds a new folder for large runtime resources to git-lfs. Unlike the large test resources, which are only accessed via a volume mount when tests are run on the Docker image, the runtime resources need to be accessible to the GATK build during the Docker build process, since they're included in the jar. AFAICT there is no `docker build` equivalent to `docker run -v`. So for now the runtime resources are git pulled into the Docker staging area, and thus onto the Docker image. We need this for @lucidtronix 's CNN branch (and possibly for @TedBrookings) if we're going to load models for resources, but longer term, we need a better solution.

### CODE REVIEW

**Comment by cmnbroad** [comment]
@jamesemery @lbergelson Any alternate suggestions ?

File: src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java [hunk 1]
```diff
@@ -270,5 +267,21 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
         Assert.assertEquals(string, IOUtils.urlDecode(encoded));
     }
 
+    @Test
+    public void testLoadLargeRuntimeResource() throws IOException {
+        final String largeResourcePath = "large/largeResourceTest.txt";
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java]: this path should be a constant somewhere

**Comment by cmnbroad** [comment]
I added a check to build.gradle to verify that the large (runtime) resources are not lfs pointer files. I also added a test to verify that there is a .git folder, and non-null ToolProvider classLoader (fixes https://github.com/broadinstitute/gatk/issues/4532). To test the pointer file check, I created a branch of this branch with a temporary commit that reverts the addition of the large resources to the Docker staging area, and [submitted](https://travis-ci.org/broadinstitute/gatk/builds/354495062) it to Travis to verify that the build fails when they're not present.

File: README.md [hunk 2]
```diff
@@ -55,6 +55,8 @@ releases of the toolkit.
 * Optional, but recommended:
     * Gradle 3.1 or greater, needed for building the GATK. We recommend using the `./gradlew` script which will
       download and use an appropriate gradle version automatically (see examples below).
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater, needed for building the GATK. Used to download the large files required to run the build and to run the test suite.
+      Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone to download the large files. The download is several hundred megabytes.
```
Comment (droazen) [inline README.md]: Maybe we should split into "build" and "runtime" requirements?

File: README.md [hunk 3]
```diff
@@ -55,6 +55,8 @@ releases of the toolkit.
 * Optional, but recommended:
     * Gradle 3.1 or greater, needed for building the GATK. We recommend using the `./gradlew` script which will
       download and use an appropriate gradle version automatically (see examples below).
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater, needed for building the GATK. Used to download the large files required to run the build and to run the test suite.
```
Comment (droazen) [inline README.md]: "run the build" -> "build GATK"

File: README.md (hunk 2 above)
Comment (droazen) [reply README.md]: Also, can you check whether `git lfs pull` still works?

File: README.md [hunk 4]
```diff
@@ -102,7 +102,7 @@ You can download and run pre-built versions of GATK4 from the following places:
         
     * This creates a zip archive in the `build/` directory with a name like `gatk-VERSION.zip` containing a complete standalone GATK distribution, including our launcher `gatk`, both the local and spark jars, and this README.    
     * You can also run GATK commands directly from the root of your git clone after running this command.
-    * Note that you *must* have a full git clone in order to build GATK. The zipped source code alone is not buildable.
+    * Note that you *must* have a full git clone in order to build GATK, incuding git-lfs files. The zipped source code alone is not buildable.
```
Comment (droazen) [inline README.md]: incuding -> including

Comment (droazen) [reply README.md]: "including the git-lfs files in `src/main/resources`" ?

File: build.gradle [hunk 5]
```diff
@@ -38,6 +38,41 @@ startScripts {
...
+            );
+        }
+    }
+}
+
+// Validate that we have a clone of the repository and that any required git-lfs files have been resolved
+task validateBuildPrerequisites {
+    if (ToolProvider.getSystemToolClassLoader() == null) {
+        throw new RuntimeException(
+                String.format(buildPrerequisitesErrorFormatString, "A (version 8) JDK must be installed."));
```
Comment (droazen) [inline build.gradle]: Extract required Java version as a named constant at the top

File: build.gradle [hunk 6]
```diff
@@ -38,6 +38,41 @@ startScripts {
...
+    }
+    if (!file(".git").isDirectory()) {
+        throw new RuntimeException(
+                String.format(buildPrerequisitesErrorFormatString,
+                        "The GATK Github repository must be cloned to run the build."));
```
Comment (droazen) [inline build.gradle]: Add "using `git clone`" 

Also provide a link to github tutorial

File: build.gradle [hunk 7]
```diff
@@ -38,6 +38,41 @@ startScripts {
...
+    }
+    // Next, ensure that the git-lfs files required to run the build are present
+    final FileTree largeResources = fileTree(dir: "src/main/resources/large")
```
Comment (droazen) [inline build.gradle]: Declare `src/main/resources/large` as a named constant at the top

File: build.gradle [hunk 8]
```diff
@@ -38,6 +38,41 @@ startScripts {
...
 }
 
+final String buildPrerequisitesErrorFormatString = "%s See https://github.com/broadinstitute/gatk#building for information on how to build GATK."
+
+// Large runtime resource files must be present at build time to be compiled into the jar, so validate
+// that the lfs files have been fetched and that the directory does not contain the git-lfs pointer files
+def validateLargeResourceFile(final File largeResourceFile, buildPrerequisitesErrorFormatString) {
+    largeResourceFile.withReader("UTF-8") { reader ->
+        final String firstLine = reader.readLine();
+        if (firstLine != null && firstLine.startsWith("version https://git-lfs.github.com/spec/v1")) {
```
Comment (droazen) [inline build.gradle]: Declare this string as a named constant somewhere

File: build.gradle [hunk 9]
```diff
@@ -38,6 +38,41 @@ startScripts {
     }
 }
 
+final String buildPrerequisitesErrorFormatString = "%s See https://github.com/broadinstitute/gatk#building for information on how to build GATK."
+
+// Large runtime resource files must be present at build time to be compiled into the jar, so validate
+// that the lfs files have been fetched and that the directory does not contain the git-lfs pointer files
+def validateLargeResourceFile(final File largeResourceFile, buildPrerequisitesErrorFormatString) {
+    largeResourceFile.withReader("UTF-8") { reader ->
+        final String firstLine = reader.readLine();
```
Comment (droazen) [inline build.gradle]: Read first X bytes instead of a full line in case the first line is very large.

File: build_docker.sh [hunk 10]
```diff
@@ -98,6 +98,11 @@ if [ -n "$STAGING_DIR" ]; then
     GIT_CHECKOUT_COMMAND="git checkout ${GITHUB_DIR}${GITHUB_TAG}"
     echo "${GIT_CHECKOUT_COMMAND}"
     ${GIT_CHECKOUT_COMMAND}
+    # pull just the large runtime resources into the staging area so they are available to the build
+    # when its done as part of the Docker build so the resources can be compiled into the jar as resources
```
Comment (droazen) [inline build_docker.sh]: its -> it's

File: build_docker.sh [hunk 11]
```diff
@@ -98,6 +98,11 @@ if [ -n "$STAGING_DIR" ]; then
...
+    GIT_PULL_LARGE_COMMAND="git lfs pull --include src/main/resources/large/"
```
Comment (droazen) [inline build_docker.sh]: Might be good to mention this command in the README for users who want to build GATK without downloading all of the large test data.

File: build_docker.sh [hunk 12]
```diff
@@ -98,6 +98,11 @@ if [ -n "$STAGING_DIR" ]; then
...
+    echo ${GIT_PULL_LARGE_COMMAND}
+    ${GIT_PULL_LARGE_COMMAND}
```
Comment (droazen) [inline build_docker.sh]: Check the exit status code here

File: src/main/java/org/broadinstitute/hellbender/utils/io/Resource.java [hunk 13]
```diff
@@ -9,6 +9,7 @@
 public final class Resource {
     private final String path;
     private final Class<?> relativeClass;
+    public final static String LARGE_RESOURCES_PATH = "large";
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/io/Resource.java]: Add a comment here explaining what this is

File: src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java [hunk 14]
```diff
@@ -270,5 +267,20 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
         Assert.assertEquals(string, IOUtils.urlDecode(encoded));
     }
 
+    @Test
+    public void testLoadLargeRuntimeResource() throws IOException {
+        final Resource largeResource = new Resource(Resource.LARGE_RESOURCES_PATH + "/largeResourceTest.txt", null);
+        final File resourceFile = IOUtils.writeTempResource(largeResource);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java]: Might want to extract a simple utility method for loading a resource and writing to a temp location (would return a File)

File: README.md (hunk 4 above)
Comment (cmnbroad) [reply README.md]: done.

File: README.md (hunk 2 above)
Comment (cmnbroad) [reply README.md]: I split this into separate sections for runtime and build. Updated to state that the lfs download is around 2GB. And `git lfs pull` does work.

File: README.md (hunk 3 above)
Comment (cmnbroad) [reply README.md]: done.

File: build.gradle (hunk 9 above)
Comment (cmnbroad) [reply build.gradle]: done.

File: build.gradle (hunk 8 above)
Comment (cmnbroad) [reply build.gradle]: done.

File: build.gradle (hunk 5 above)
Comment (cmnbroad) [reply build.gradle]: Done.

File: build.gradle (hunk 7 above)
Comment (cmnbroad) [reply build.gradle]: Done.

File: build.gradle (hunk 6 above)
Comment (cmnbroad) [reply build.gradle]: These are all switched to use string interpolation now. The embedded buildPrerequisites message directs the user to the gatk github repo.

File: build_docker.sh (hunk 10 above)
Comment (cmnbroad) [reply build_docker.sh]: done.

File: build_docker.sh (hunk 11 above)
Comment (cmnbroad) [reply build_docker.sh]: Done.

File: src/main/java/org/broadinstitute/hellbender/utils/io/Resource.java (hunk 13 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/utils/io/Resource.java]: Done.

**Comment by cmnbroad** [comment]
@droazen please take a look at the updated README.md in this PR.

File: src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java [hunk 15]
```diff
@@ -229,6 +229,20 @@ public static File writeTempResource(Resource resource) {
         return temp;
     }
 
+    /**
+     * Create a resource from a path and a relative class, and write it to a temporary file.
+     * @param resourcePath Relative or absolute path to the class.
+     * @param relativeClass Relative class to use as a class loader and for a relative package.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java]: Javadoc `@param` tags usually come at the end, rather than in the middle of the comment text.

File: src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java [hunk 16]
```diff
@@ -270,5 +267,42 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
...
+
+        String resourceContents = "";
+        try (final FileReader fr = new FileReader(resourceFile);
+             final BufferedReader br = new BufferedReader(fr)) {
+            resourceContents = br.readLine();
+        }
+        finally {
+            resourceFile.delete();
+        }
+        Assert.assertEquals(resourceContents, "line in large resource file");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java]: Declare expected text as a constant, and use it in both test methods.

File: src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java [hunk 17]
```diff
@@ -270,5 +267,42 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
         Assert.assertEquals(string, IOUtils.urlDecode(encoded));
     }
 
+    @DataProvider(name="resourcePaths")
+    public Object[][] getResourcePaths() {
+        return new Object[][] {
+                { Resource.LARGE_RUNTIME_RESOURCES_PATH + "/largeResourceTest.txt", null }
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java]: Can you add a test case involving a non-null relative class as well?

File: README.md [hunk 18]
```diff
@@ -50,28 +50,36 @@ releases of the toolkit.
...
-    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater (needed to download large files for the complete test suite).
-      Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone to download the large files. The download is several hundred megabytes.
-* Alternatively, pre-packaged images with all needed dependencies installed can be found on [our dockerhub repository](https://hub.docker.com/r/broadinstitute/gatk/). This requires a recent version of the docker client, which can be found on the [docker website](https://www.docker.com/get-docker).
+* To build GATK:
+    * A Java 8 JDK
+    * Git 2.5 or greater
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater. Required to download the large files used to build GATK and
+      run the test suite. Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone
+      to download all of the large files. The full download is approximately 2 gigabytes. Alternatively, to pull the minimal
+      set of large `lfs` resource files required to only build ()without running tests), the test resources can be excluded
```
Comment (droazen) [inline README.md]: `()without running tests)` -> `(without running tests)`

File: README.md [hunk 19]
```diff
@@ -50,28 +50,36 @@ releases of the toolkit.
...
+      from the download by running `git lfs pull --include src/main/resources/large/` instead of `git lfs pull`. This greatly
```
Comment (droazen) [inline README.md]: Can we modify `build.gradle` to run this command automatically at build time, if lfs stubs are detected? Let's do that (if possible) -- then this set of instructions can be simplified to just "install git lfs"

File: README.md [hunk 20]
```diff
@@ -50,28 +50,36 @@ releases of the toolkit.
...
       (developers should run ```./gradlew createPythonPackageArchive```, followed by
       ```conda env create -n gatk -f scripts/gatkcondaenv.yml``` from within the root of the repository clone).
       To activate the environment once it has been created, run the command ```source activate gatk```. See the
       [Conda](https://conda.io/docs/user-guide/tasks/manage-environments.html) documentation for
       additional information about using and managing Conda environments.
     * R 3.2.5 (needed for producing plots in certain tools, and for running the test suite)
-    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater (needed to download large files for the complete test suite).
-      Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone to download the large files. The download is several hundred megabytes.
-* Alternatively, pre-packaged images with all needed dependencies installed can be found on [our dockerhub repository](https://hub.docker.com/r/broadinstitute/gatk/). This requires a recent version of the docker client, which can be found on the [docker website](https://www.docker.com/get-docker).
+* To build GATK:
```
Comment (droazen) [inline README.md]: Is R a build-time dependency as well?

File: build.gradle [hunk 21]
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
...
+                "$largeResourceFile appears to be a git-lfs pointer file." +
+                        " The resource files in src/main/resources/large must be downloaded using git-lfs. " +
+                        " $buildPrerequisitesMessage")
+    }
+}
+
+// Validate that we have a clone of the repository and that any required git-lfs files have been resolved
+def validateBuildPrerequisites(javaVersion, largeResourcesFolder, lfsSentinel, buildPrerequisitesMessage) {
+    if (ToolProvider.getSystemToolClassLoader() == null) {
+        throw new GradleException("The SystemToolClassLoader obtained from the ToolProvider is null. A Java $RequiredJavaVersion JDK must be installed. $buildPrerequisitesMessage")
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: I think it happens when you have a JRE rather than a JDK (see https://github.com/broadinstitute/gatk/issues/4532). Also, although I haven't tried it I suspect it will happen with JDK9.

File: README.md (hunk 20 above)
Comment (cmnbroad) [reply README.md]: Only to run tests, AFAIK. I added it to this section as well saying its needed for running tests.

File: README.md (hunk 18 above)
Comment (cmnbroad) [reply README.md]: Done.

File: README.md (hunk 19 above)
Comment (cmnbroad) [reply README.md]: Done. I changed the build to check for stubs, and to only run `git lfs pull --include src/main/resources/large/` if it finds any, then validate again. Also, we still have to manually download the files into the staging area, since the Docker doesn't have git-lfs installed.

File: src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java (hunk 15 above)
Comment (cmnbroad) [reply src/main/java/org/broadinstitute/hellbender/utils/io/IOUtils.java]: Doh.

File: build.gradle [hunk 22]
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
 
 final baseJarName = 'gatk'
 final secondaryBaseJarName = 'hellbender'
-
 final pythonPackageArchiveName = 'gatkPythonPackageArchive.zip'
+final largeResourcesFolder = "src/main/resources/large"
+final lfsFileSentinel = "version https://git-lfs.github.com/spec/v1"
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: I changed all of these names to use "stub", but I'm not sure thats better, since AFACIT git-lfs refers to these as pointer files in all of the help and doc.

File: build.gradle [hunk 23]
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
...
+def validateLargeResourceFile(largeResourceFile, lfsSentinel, buildPrerequisitesMessage) {
+    def readBytesFromFile = { largeFile, n ->
+        final byte[] bytes = new byte[n]
+        largeResourceFile.withInputStream { stream -> stream.read(bytes, 0, bytes.length) }
+        return bytes
+    }
+    final byte[] actualBytes = readBytesFromFile(largeResourceFile, lfsSentinel.length());
+    if (new String(actualBytes, "UTF-8") == lfsSentinel) {
+        throw new GradleException(
+                "$largeResourceFile appears to be a git-lfs pointer file." +
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: Done.

File: build.gradle [hunk 24]
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
...
+                        " The resource files in src/main/resources/large must be downloaded using git-lfs. " +
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: Done.

File: build.gradle [hunk 25]
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
...
 final baseJarName = 'gatk'
 final secondaryBaseJarName = 'hellbender'
-
 final pythonPackageArchiveName = 'gatkPythonPackageArchive.zip'
+final largeResourcesFolder = "src/main/resources/large"
+final lfsFileSentinel = "version https://git-lfs.github.com/spec/v1"
+final buildPrerequisitesMessage = "See https://github.com/broadinstitute/gatk#building for information on how to build GATK."
+
+// Large runtime resource files must be present at build time to be compiled into the jar, so validate
+// that the lfs files have been fetched and that the directory does not contain the git-lfs pointer files
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: Done.

File: build.gradle [hunk 26]
```diff
@@ -66,8 +67,42 @@ final testNGVersion = '6.11'
...
+// Validate that we have a clone of the repository and that any required git-lfs files have been resolved
+def validateBuildPrerequisites(javaVersion, largeResourcesFolder, lfsSentinel, buildPrerequisitesMessage) {
+    if (ToolProvider.getSystemToolClassLoader() == null) {
+        throw new GradleException("The SystemToolClassLoader obtained from the ToolProvider is null. A Java $RequiredJavaVersion JDK must be installed. $buildPrerequisitesMessage")
+    }
+    if (!file(".git").isDirectory()) {
+        throw new GradleException("The GATK Github repository must be cloned using \"git clone\" to run the build. $buildPrerequisitesMessage")
+    }
+    // Next, ensure that the git-lfs files required to run the build are present
+    def largeResources = fileTree(dir: largeResourcesFolder)
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: Only files.

File: README.md (hunk 18 above)
Comment (cmnbroad) [reply README.md]: Done.

File: README.md (hunk 19 above)
Comment (cmnbroad) [reply README.md]: Done see above note.

**Comment by cmnbroad** [comment]
Back to @droazen. Responded to all of the comments. I did an interim commit to test the lfs changes on docker/travis so there are two commits for this round.

File: src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java [hunk 27]
```diff
@@ -270,5 +267,44 @@ public void testUrlEncodeDecode(final String string, final String encoded) {
...
+        };
+    }
+
+    @Test(dataProvider = "resourcePaths")
+    public void testWriteTempResource(
+            final String resourcePath, final Class<?> relativeClass, final String expectedFirstLine) throws IOException
+    {
+        final Resource largeResource = new Resource(resourcePath, relativeClass);
+        final File resourceFile = IOUtils.writeTempResource(largeResource);
+        final String resourceContentsFirstLine = getFirstLineAndDeleteTempFile(resourceFile);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java]: Should the `IOUtils.writeTempResource()` methods set the extracted resource to be deleted on JVM exit? It seems like we will end up leaving garbage around if we don't do that in the methods themselves.

File: README.md [hunk 28]
```diff
@@ -50,28 +50,37 @@ releases of the toolkit.
...
-      Run `git lfs install` after downloading, followed by `git lfs pull` from the root of your git clone to download the large files. The download is several hundred megabytes.
-* Alternatively, pre-packaged images with all needed dependencies installed can be found on [our dockerhub repository](https://hub.docker.com/r/broadinstitute/gatk/). This requires a recent version of the docker client, which can be found on the [docker website](https://www.docker.com/get-docker).
+    * R 3.2.5 (needed for producing plots in certain tools)
+* To build GATK:
+    * A Java 8 JDK
+    * Git 2.5 or greater
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater. Required to download the large files used to build GATK, and
+      test files required to run the test suite. Run `git lfs install` after downloading, followed by `git lfs pull` from
+      the root of your git clone to download all of the large files, including those required to run the test suite. The
+      full download is approximately 2 gigabytes. Alternatively, you can skip this step since the build will itself use
```
Comment (droazen) [inline README.md]: "Alternatively, " -> "Alternatively, if you are just building GATK and not running the test suite, "

File: build.gradle [hunk 29]
```diff
@@ -66,8 +67,66 @@ final testNGVersion = '6.11'
...
+
+// Large runtime resource files must be present at build time to be compiled into the jar, so force
+// the contents of the large runtime resources to be resolved to real files rather than stub files
+def gitPullLargeResources(largeResourcesFolder) {
+    final String execLFSPull = "git lfs pull --include $largeResourcesFolder";
+    try {
+        println "Fetching large runtime resources via: $execLFSPull"
+        def retCode = execLFSPull.execute().waitFor()
+        if (retCode.intValue() != 0) {
+            throw new GradleException("git-lfs pull execution failed: $retCode");
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: Done.

File: build.gradle [hunk 30]
```diff
@@ -66,8 +67,66 @@ final testNGVersion = '6.11'
...
+        }
+        return retCode
+    } catch (IOException e) {
+        throw new GradleException("git-lfs must be installed to run the GATK build", e)
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: I can't be sure. I changed the message to report the IOException and suggest checking git-lfs.

File: build.gradle [hunk 31]
```diff
@@ -66,8 +67,66 @@ final testNGVersion = '6.11'
...
+        throw new GradleException("The GATK Github repository must be cloned using \"git clone\" to run the build. $buildPrerequisitesMessage")
+    }
+    // Next, ensure that the git-lfs files required to run the build are present
+    if (checkLargeResourcesForStubFiles(largeResourcesFolder, lfsStubFileHeader, buildPrerequisitesMessage)) {
+        // try to pull once, then verify again
+        gitPullLargeResources(largeResourcesFolder)
+        if (checkLargeResourcesForStubFiles(largeResourcesFolder, lfsStubFileHeader, buildPrerequisitesMessage)) {
+            throw new GradleException(
+                    "$largeResourcesFolder contains git-lfs stub files." +
+                            " The resource files in $largeResourcesFolder must be downloaded using git-lfs. " +
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: Done.

File: build.gradle [hunk 32]
```diff
@@ -66,8 +67,66 @@ final testNGVersion = '6.11'
...
+    }
+    def largeResourceFiles = fileTree(dir: largeResourcesFolder)
+    return largeResourceFiles.any() { f ->
+        final byte[] actualBytes = readBytesFromFile(f, lfsStubFileHeader.length());
+        return new String(actualBytes, "UTF-8") == lfsStubFileHeader
+    }
+}
+
+// Validate that we have a clone of the repository and that any required git-lfs files have been resolved
+def validateBuildPrerequisites(javaVersion, largeResourcesFolder, lfsStubFileHeader, buildPrerequisitesMessage) {
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: done.

File: build.gradle [hunk 33]
```diff
@@ -262,6 +321,8 @@ processResources {
 processTestResources {
     //Don't waste time packaging unnecessary test data into the test resources:
     include "org/broadinstitute/hellbender/utils/config/*"
+    //Required for IOUtils resource tests
+    include "org/broadinstitute/hellbender/utils/io/*"
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply build.gradle]: Only if you add a resource that needs to be packaged into the test jar as part of the test resources.

File: src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java (hunk 27 above)
Comment (cmnbroad) [reply src/test/java/org/broadinstitute/hellbender/utils/io/IOUtilsUnitTest.java]: We probably should. I'm going to do that in a separate PR though since all of the call sites should be cleaned up as well.

File: README.md [hunk 34]
```diff
@@ -50,28 +50,38 @@ releases of the toolkit.
...
+    * R 3.2.5 (needed for producing plots in certain tools)
+* To build GATK:
+    * A Java 8 JDK
+    * Git 2.5 or greater
+    * [git-lfs](https://git-lfs.github.com/) 1.1.0 or greater. Required to download the large files used to build GATK, and
+      test files required to run the test suite. Run `git lfs install` after downloading, followed by `git lfs pull` from
+      the root of your git clone to download all of the large files, including those required to run the test suite. The
+      full download is approximately 2 gigabytes. Alternatively, if you are just building GATK and not running the test
+      suite, you can skip this step since the build itself will use git-lfs to download the minimal set of large `lfs`
+      resource files required to complete the build. The tests resources will not be downloaded, but this greatly reduces
```
Comment (droazen) [inline README.md]: "tests resources" -> "test resources"

File: build.gradle [hunk 35]
```diff
@@ -68,6 +69,72 @@ final baseJarName = 'gatk'
...
 final pythonPackageArchiveName = 'gatkPythonPackageArchive.zip'
+final largeResourcesFolder = "src/main/resources/large"
+final buildPrerequisitesMessage = "See https://github.com/broadinstitute/gatk#building for information on how to build GATK."
+
+// Returns true if any files in the target folder are git-lfs stub files.
+def checkForLFSStubFiles(targetFolder) {
+    final lfsStubFileHeader = "version https://git-lfs.github.com/spec/v1"  // first line of a git-lfs stub file
+    def readBytesFromFile = { largeFile, n ->
+        final byte[] bytes = new byte[n]
+        largeFile.withInputStream { stream -> stream.read(bytes, 0, bytes.length) }
```
Comment (droazen) [inline build.gradle]: Does the input stream here get closed?

Comment (cmnbroad) [reply build.gradle]: Yes, its equivalent to try-wth-resources.

File: README.md (hunk 34 above)
Comment (cmnbroad) [reply README.md]: Done.
